WIP: Convert the UI to be React based

wbushey opened the pull request.

wbushey committed:
Adding react_rails

wbushey committed:
First React component

wbushey committed:
Added Jasmine, and first React spec

wbushey committed:
Using jquery-rails version of jQuery
Using jquery-ui-rails

wbushey committed:
Upgrading bootstrap.js to 2.3.2

wbushey committed:
jQuery 1.8 -> 1.11: live -> on

wbushey committed:
More development of the Field component
Finished display of Field

wbushey committed:
Text fields update their values

wbushey committed:
Refactored Field into several focused components

wbushey committed:
All components can receive input

wbushey committed:
Fields can have initial values set

wbushey committed:
Fields can display errors

wbushey committed:
Starting the Edit Profile view in React
Readding connection pool dependency

wbushey committed:
Working shipping options

wbushey committed:
Created Survey form partial
DRYed up form partials
Added a couple of missing translations
Fields pass on required and private attributes
A few display tweaks

wbushey committed:
Removing unneeded server side partials
Rendering Edit Profile via React

wbushey committed:
Replaced display of location search with React component
Named partials partials

wbushey committed:
Implemented LocationSearchView
Moved most map related JS into a ThingMap class
Add spec helper to mock out Google Maps library during tests

wbushey committed:
Old work in some state, including some failing tests

Commit:
Updating .gitignore and setting up working environment

Commit:
Fixes #23

wbushey committed:
Fixed one thing about a still failing spec

wbushey committed:
Component Fields can take onStateChange callbacks (#35)

* All field components accept a callback function for state change
* Addressing warnings about keys and proptypes
Fixes #33

jhsu802701 commented:
I'd like to provide input, but I need to get up to speed on a few things.  I'm not familiar with React.  Why do we want a React-based UI?  What does it replace, and why is React better?  How should the conflict in app/views/layouts/application.html.haml be resolved?  In addition to running all of the tests and making sure that they pass, is there anything else I need to check?

SIDE NOTE: The large number of changes makes this pull request intimidating to review.  I prefer to make numerous small changes instead of one big one.  Making too many changes all at once makes troubleshooting much more difficult if something doesn't pan out.

wbushey commented:
Good questions @jhsu802701. The whats and whys are covered a bit by the description of the Rebuild UI in React project (https://github.com/OpenTwinCities/adopt-a-tree/projects/2 and click 'Details'). Basically, this project aims to re-write the UI to be much more testable and modular, with the ultimate goal of making future UI development easier. 

As far as testing, this branch includes JavaScript unit tests in spec/javascripts. At the moment, running these in the command line will error out, but they can be run in browser by starting the rails server as normal and visiting <server_address>/specs .

As a rule of thumb, I definitely agree with you that MRs should be much smaller. Once work started on this branch, it became clear that this project is a pretty significant undertaking, and that it won't make sense to merge it into master until quite a bit of work has been done. So, we decided to treat this branch as a long term project branch. Folks wishing to work on this project should create merge requests against this branch (`react`), not master. When this branch is ready to be deployed, we'll then merge it into master.

Another good question regarding conflicts. At the moment, I'm not sure (I know, that's a bad answer). I just rebased this branch with master and addressed a couple of conflicts in doing that; I'm actually a little surprised that GitHub is still showing conflicts.

wbushey committed:
Finishing update of gems from master

wbushey committed:
Merge branch 'master' into react

wbushey committed:
Fixing Gemfile
